### DISCUSSION TITLE
Extend `getStaticProps` to API Routes

### DISCUSSION THREAD

**Discussion by ctjlewis** [vercel/next.js#25969] [Ideas]
### Describe the feature you'd like to request

I have a very expensive process located at `api/some-endpoint.ts` and only want to revalidate the output once every `t` seconds. We can do this easily with `getStaticProps()` for components, but not for API Routes.

However, since they also follow the `export default` pattern, it seems like the same approach could be taken for API responses.

### Describe the solution you'd like

The most appropriate way to do this IMO is to extend `getStaticProps()` to API Routes and inject into the function by changing:

https://github.com/vercel/next.js/blob/e5d0a30a2eda1af11307a1cc123cba69cb2d6a90/packages/next/next-server/server/next-server.ts#L1163

To:

```ts
const staticProps = getStaticProps();
await pageModule.default(req, res, ...staticProps.props)
```

And extend the revalidate logic (e.g. `setRevalidateHeaders` in `next-server/send-payload`). But because API response rendering (e.g. `handleApiRequest`) 

### Describe alternatives you've considered

I am working around this by manually setting the `Cache-Control` header into my API Route, but it seems like development mode will always force a `no-cache` so `next start` is necessary to test the override.

**Comment by ctjlewis** [comment]
Struggling to work around and manually override Cache-Control, even with `next.config.js`:

```js
module.exports = {
  async headers() {
    return [
      {
        source: '/api/[endpoint]',
        headers: [
          {
            key: 'Cache-Control',
            value: 'max-age=900',
          },
        ],
      },
    ];
  },
}
```

**Comment by ctjlewis** [comment]
In my API Route, I was able to get this working with:

```ts
res.setHeader(
  'Cache-Control',
  `public, max-age=${15 * 60}, stale-while-revalidate`
);
```

But this cannot be tested with local Next Server apparently, and even with `next start`, the `Cache-Control` property is not appropriately mocked. I fiddled for a few hours before thinking to deploy to the actual CDN and make sure it was actually working (despite failing in local testing, even in production).

**Comment by GorlikItsMe** [comment]
For 4 hours I have been wondering why the getStaticProps does not work with api enpoints.
it would be great if this feature will be added. Also good idea is add to docs information "this dont work with api routes" or something
